src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java

```diff
@@ -0,0 +1,311 @@
+                "withCredentials([usernamePassword(credentialsId: 'test', usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')]) {\n"+
+                "  echo \"${env.USERNAME} ${env.PASSWORD}\"\n"+
+                "}"
+        ));
+        WorkflowRun run  = job.scheduleBuild2(0).getStartCondition().get();
+        r.waitForCompletion(run);
+        FlowExecution exec = run.getExecution();
+        String log = r.getLog(run);
+        ForkScanner scanner = new ForkScanner();
+        List<FlowNode> filtered = scanner.filteredNodes(exec, new DescriptorMatchPredicate(BindingStep.DescriptorImpl.class));
+
+        // Check the binding step is OK
+        Assert.assertEquals(8, filtered.size());
+        FlowNode node = Collections2.filter(filtered, FlowScanningUtils.hasActionPredicate(ArgumentsActionImpl.class)).iterator().next();
+        ArgumentsActionImpl act = node.getPersistentAction(ArgumentsActionImpl.class);
+        Assert.assertNotNull(act.getArgumentValue("bindings"));
+        Assert.assertNotNull(act.getArguments().get("bindings"));
+
+        // Test that masking really does mask bound credentials appropriately
+        filtered = scanner.filteredNodes(exec, new DescriptorMatchPredicate(EchoStep.DescriptorImpl.class));
+        for (FlowNode f : filtered) {
+            act = f.getPersistentAction(ArgumentsActionImpl.class);
+            Assert.assertEquals(ArgumentsAction.NotStoredReason.MASKED_VALUE, act.getArguments().get("message"));
+            Assert.assertNull(ArgumentsAction.getStepArgumentsAsString(f));
+        }
+
+        List<FlowNode> allStepped = scanner.filteredNodes(run.getExecution().getCurrentHeads(), FlowScanningUtils.hasActionPredicate(ArgumentsActionImpl.class));
+        Assert.assertEquals(6, allStepped.size());  // One ArgumentsActionImpl per block or atomic step
+
+        testDeserialize(exec);
+    }
+
+    @Test
+    public void simpleSemaphoreStep() throws Exception {
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "p");
+        job.setDefinition(new CpsFlowDefinition(
+                "semaphore 'wait'"
+        ));
+        WorkflowRun run  = job.scheduleBuild2(0).getStartCondition().get();
+        SemaphoreStep.waitForStart("wait/1", run);
+        FlowNode semaphoreNode = run.getExecution().getCurrentHeads().get(0);
+        CpsThread thread = CpsThread.current();
+        SemaphoreStep.success("wait/1", null);
+        r.waitForCompletion(run);
+        testDeserialize(run.getExecution());
+    }
+
+    @Test
+    public void testArgumentDescriptions() throws Exception {
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "argumentDescription");
+        job.setDefinition(new CpsFlowDefinition(
+                "echo 'test' \n " +
+                        " node('master') { \n" +
+                        "   retry(3) {\n"+
+                        "     if (isUnix()) { \n" +
+                        "       sh 'whoami' \n" +
```

Comment:
Is there a particular reason why this matters?  Something specific that is *explicitly wrong* with it, not just something that you prefer to do differently?